src/System.Sorting/tests/SortTests.cs

```diff
@@ -0,0 +1,140 @@
+        {
+
+
+            T[] control = SimpleSort<T>(array, index, length, comparer);
+
+            //TODO - Can use Reflection to get the derivedTypes of Sorting and invoking the methods. 
+            //It's not so trivial because we are using Reflection with Generic <T>. 
+            //TypeInfo[] mySortTestAlgorithms = FindAllDerivedTypesOfSorting();
+
+            Sorting[] sortAlgos = new Sorting[] { new IntroSorting(), new InsertionSorting() };
+            foreach(var algo in sortAlgos)
+            {
+                T[] spawn = (T[])(array.Clone());
+                // get instance from algo
+                algo.Sort<T>(spawn, index, length, comparer);
+                Assert.True(ArraysAreEqual<T>((T[])spawn, control, comparer));
+            }
+
+         }
+
+        private static T[] SimpleSort<T>(T[] a, int index, int length, IComparer<T> comparer)
+        {
+            T[] result = (T[])(a.Clone());
+            if (length < 2)
+                return result;
+
+            for (int i = index; i < index + length - 1; i++)
+            {
+                T tmp = result[i];
+                for (int j = i + 1; j < index + length; j++)
+                {
+                    if (comparer.Compare(tmp, result[j]) > 0)
+                    {
+                        result[i] = result[j];
+                        result[j] = tmp;
+                        tmp = result[i];
+                    }
+                }
+            }
+            return result;
+        }
+
+        private static bool ArraysAreEqual<T>(T[] a, T[] b, IComparer<T> comparer)
+        {
+            // If the same instances were passed, this is unlikely what the test intended.
+            Assert.False(Object.ReferenceEquals(a, b));
+
+            if (a.Length != b.Length)
+                return false;
+            for (int i = 0; i < a.Length; i++)
+            {
+                if (0 != comparer.Compare(a[i], b[i]))
+                    return false;
+            }
+            return true;
+        }
+
+        private class IntegerComparer : IComparer<int>
+        {
+            public int Compare(object x, object y)
+            {
+                return Compare((int)x, (int)y);
+            }
+
+            public int Compare(int x, int y)
+            {
+                return x - y;
+            }
+        }
+
+        private class StringComparer : IComparer<string>
+        {
+            public int Compare(object x, object y)
+            {
+                return Compare((string)x, (string)y);
+            }
+
+            public int Compare(string x, string y)
+            {
+                if (x == y)
+                    return 0;
+                if (x == null)
+                    return -1;
+                if (y == null)
+                    return 1;
+                return x.CompareTo(y);
+            }
+        }
+
+        public static  TypeInfo [] FindAllDerivedTypesOfSorting()
```

Comment:
I think this is adding too much magic (and it's difficult to make it work as you wanted).  What you could consider doing is to make the test class itself (with the Facts/Theories) abstract, with an abstract get property of type `Sorting`, which you use in the Facts and Theories to actually get the sorter.  Then have two public implementations of the class, which just returns a cached IntroSorting or InsertionSorting instance for that abstract property.  XUnit should be able to see these public instances and run the tests.  We do this some other places in CoreFX and it works out pretty well.